flink-java/src/main/java/org/apache/flink/api/java/utils/PythonDependencyManager.java

```
@@ -0,0 +1,282 @@
+ * cache.
+ */
+@Internal
+public class PythonDependencyManager {
+
+	public static final String PYTHON_FILE_PREFIX = "python_file";
+	public static final String PYTHON_REQUIREMENTS_FILE_PREFIX = "python_requirements_file";
+	public static final String PYTHON_REQUIREMENTS_CACHE_PREFIX = "python_requirements_cache";
+	public static final String PYTHON_ARCHIVE_PREFIX = "python_archive";
+	public static final String CLUSTER = "cluster";
+	public static final String FILE = "file";
+	public static final String CACHE = "cache";
+	public static final String FILE_DELIMITER = ",";
+	public static final String PARAM_DELIMITER = "#";
+
+	// Metadata of Python dependencies.
+
+	public static final ConfigOption<Map<String, String>> PYTHON_FILES =
+		ConfigOptions.key("python.internal.files-key-map").mapType().noDefaultValue();
+	public static final ConfigOption<Map<String, String>> PYTHON_REQUIREMENTS_FILE =
+		ConfigOptions.key("python.internal.requirements-file-key").mapType().noDefaultValue();
+	public static final ConfigOption<Map<String, String>> PYTHON_ARCHIVES =
+		ConfigOptions.key("python.internal.archives-key-map").mapType().noDefaultValue();
+
+	private final Configuration config;
+	private final List<Tuple2<String, DistributedCache.DistributedCacheEntry>> cachedFiles;
+	private int counter = -1;
+
+	public PythonDependencyManager(
+		Configuration config,
+		List<Tuple2<String, DistributedCache.DistributedCacheEntry>> cachedFiles) {
+		this.config = Preconditions.checkNotNull(config);
+		this.cachedFiles = cachedFiles;
+	}
+
+	/**
+	 * Adds a Python dependency which could be .py files, Python packages(.zip, .egg etc.) or
+	 * local directories. The dependencies will be added to the PYTHONPATH of the Python UDF worker
+	 * and the local Py4J python client.
+	 *
+	 * @param filePath The path of the Python dependency.
+	 */
+	public void addPythonFile(String filePath) {
+		Preconditions.checkNotNull(filePath);
+		if (fileAlreadyRegistered(filePath)) {
```

Should we throw an exception if this case occurs as usually this is not expected?

---

This case is expected because the `configure()` method will be called when the 
`translate()` method of the Table API planners is called. As the `translate()` can be called many times in interactive programing, a python file will be registered multiple times.